### PR TITLE
Detect and handle rollbacks on the RPC nodes

### PR DESCRIPTION
**Pull request by Stefan-Ethernal**
Addresses #667 for the RPC nodes.

This PR checks for block hashes and batch numbers mismatch during batches stage and triggers rollback immediately if inconsistencies are detected. 

## Verification steps
This could be tested using a Kurtosis, with one RPC node + another sequencer outside of kurtosis. 

Here are roughly the steps:
1. start a kurtosis cdk stack as usual
`kurtosis run --enclave cdk-v1 --args-file params.yml .`
Then, we need to run two cdk-erigon nodes, one as a sequencer that creates a new fork, and another one as an RPC node who syncs initially from the sequencer within kurtosis, and then switched off to the new sequencer that runs on its own fork.
2. collect the config files from kurtosis to local, for running the local cdk-erigon nodes:
```bash
e9dd04e13924   cdk-erigon-node-chain-allocs
8bcc8f21261a   cdk-erigon-node-chain-config
0fe6dadd740d   cdk-erigon-node-chain-spec-artifact
64eaece75701   cdk-erigon-node-chain-spec-artifact-sequencer
4fdf83e16899   cdk-erigon-node-config-artifact
c0b4d9fd0114   cdk-erigon-node-config-artifact-sequencer
```
Here is a dumb script to do so:
```bash
#!/bin/bash

# Create directories
mkdir -p rpc sequencer

# Download files
kurtosis files download cdk-v1 cdk-erigon-node-chain-allocs
kurtosis files download cdk-v1 cdk-erigon-node-chain-config
kurtosis files download cdk-v1 cdk-erigon-node-chain-spec-artifact
kurtosis files download cdk-v1 cdk-erigon-node-config-artifact
kurtosis files download cdk-v1 cdk-erigon-node-config-artifact-sequencer

# Move files to RPC folder
cp cdk-erigon-node-chain-allocs/* rpc/
cp cdk-erigon-node-chain-config/* rpc/
cp cdk-erigon-node-chain-spec-artifact/* rpc/
cp cdk-erigon-node-config-artifact/* rpc/

# Move files to Sequencer folder
cp cdk-erigon-node-chain-allocs/* sequencer/
cp cdk-erigon-node-chain-config/* sequencer/
cp cdk-erigon-node-chain-spec-artifact/* sequencer/
cp cdk-erigon-node-config-artifact-sequencer/* sequencer/

rm -rf cdk-erigon-node-chain-allocs
rm -rf cdk-erigon-node-chain-config
rm -rf cdk-erigon-node-chain-spec-artifact
rm -rf cdk-erigon-node-config-artifact
rm -rf cdk-erigon-node-config-artifact-sequencer
echo "Files have been downloaded and organized into rpc and sequencer folders."
```
3. adapt the configuration in the config.yaml for rpc:
```
zkevm.l2-sequencer-rpc-url -> output of $(kurtosis port print cdk-v1 cdk-erigon-sequencer-001 rpc)
zkevm.l2-datastreamer-url -> output of $(kurtosis port print cdk-v1 cdk-erigon-sequencer-001 data-streamer)
zkevm.l1-rpc-url -> output of $(kurtosis port print cdk-v1 el-1-geth-lighthouse rpc)
```
and sequencer:
```
zkevm.l1-rpc-url -> output of $(kurtosis port print cdk-v1 el-1-geth-lighthouse rpc)
zkevm.l1-sync-start-block: 1 # this flag will enable L1 recovery mode
```
Notice that you might need to change the `datadir` and service ports and assign unique values to them (`pprof.port`, `http.port`, `torrent.port`, `metrics.port`), otherwise both sequencer and rpc running on your local machine will have the same port and they will collide.
4. With configs modified, we can start local sequencer and rpc node.
Initially, we want to start sequencer in recovery mode:
```
CDK_ERIGON_SEQUENCER=1 build/bin/cdk-erigon  --config="path/to/sequencer/config.yaml" --maxpeers 0
```
It will sync from L1 in kurtosis and stay synced with kurtosis.
5. Then start RPC:
```
build/bin/cdk-erigon  --config="path/to/rpc/config.yaml" --maxpeers 0
```
6. After both nodes are in synced, we will prepare the fork.
Stop the local sequencer, delete this param `zkevm.l1-sync-start-block` from the sequencer config, and restart the sequencer. Now, the sequencer should be mining on its own fork.
7. Wait for a minute to make sure RPC node have synced certain number of blocks that will be re-sequenced.
8. To test the changes, stop RPC node, and change the data stream endpoint from the sequencer running in kurtosis to the one running locally, most likely it will be `zkevm.l2-datastreamer-url: 127.0.0.1:6900` . Then restart RPC node.
9. At this point, if everything works as expected, your RPC node should detect the fork and start rolling back.

### CODE REVIEW

File: zk/stages/stage_batches.go [hunk 1]
```diff
@@ -150,7 +152,7 @@ func SpawnStageBatches(
 
 	highestVerifiedBatch, err := stages.GetStageProgress(tx, stages.L1VerificationsBatchNo)
 	if err != nil {
-		return fmt.Errorf("could not retrieve l1 verifications batch no progress")
+		return errors.New("could not retrieve l1 verifications batch no progress")
```
Comment (cffls) [inline zk/stages/stage_batches.go]: Is this change required in this PR?

File: zk/stages/stage_batches.go [hunk 2]
```diff
@@ -890,3 +906,54 @@ func writeL2Block(eriDb ErigonDb, hermezDb HermezDb, l2Block *types.FullL2Block,
 
 	return nil
 }
+
+// findAncestorBlockNumber searches the latest ancestor block from the database, that matches specified hash using the binary search.
+// Parameters:
+//   - db: An instance of ErigonDb, which provides methods to read data from the blockchain database.
+//   - dsClient: An instance of DatastreamClient, that provides mechanism to read data from the sequencer.
+//   - latestBlockNum: The upper bound of search interval.
+//   - targetHash: The hash of the block that the function is trying to find.
```
Comment (cffls) [inline zk/stages/stage_batches.go]: Since this parameter is only used in logging, we can remove it from this function and just print it in the caller of this function if necessary.

File: zk/stages/stage_batches.go [hunk 3]
```diff
@@ -890,3 +906,54 @@ func writeL2Block(eriDb ErigonDb, hermezDb HermezDb, l2Block *types.FullL2Block,
...
+		if err != nil {
+			return 0, err
+		}
+
+		midBlockDbHash, err := db.ReadCanonicalHash(midBlockNum)
+		if err != nil {
+			return 0, err
+		}
+
+		if midBlockStream.L2Blockhash == midBlockDbHash {
```
Comment (cffls) [inline zk/stages/stage_batches.go]: It occurs to me that in a resequence event, a block's parent hash can stay the same, but the batch number of the block could be different. Could you please add a condition check to verify the batch number of a block also matches the datastream? The batch number stored for local block could be fetched by `hermezDb.GetBatchNoByL2Block(blockNo)`.

Comment (cffls) [reply zk/stages/stage_batches.go]: With the addition of checking batch number, we need to also check if the batch number of an already processed block matches to the data stream [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/92e7d22534bc103203a4a699b8de60a4a5df4c3e/zk/stages/stage_batches.go#L287-L291).

**Review by cffls — COMMENTED** [review]
Looks good overall. Had a few comments on the details. Could you please also add some unit tests? Thanks!

File: zk/stages/stage_batches.go (hunk 1 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Certainly not. I just wanted to leave the code in better shape than I found it and fix a couple of warnings. :slightly_smiling_face: I have dropped the commit that fixes those warnings, so the diff is focused..

File: zk/stages/stage_batches.go (hunk 2 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Good point! Fixed (removed the parameter, we don't need it anymore) https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/ea58d697ed124dc75afaaf9659f515582dff7f0f

File: zk/stages/stage_batches.go (hunk 3 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: > It occurs to me that in a resequence event, a block's parent hash can stay the same, but the batch number of the block could be different. Could you please add a condition check to verify the batch number of a block also matches the datastream? The batch number stored for local block could be fetched by hermezDb.GetBatchNoByL2Block(blockNo).

Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/ea58d697ed124dc75afaaf9659f515582dff7f0f.

**Comment by mandrigin** [comment]
Yeah, overall looks good but we need tests.

**Comment by Stefan-Ethernal** [comment]
> Yeah, overall looks good but we need tests.

It is work in progress (that's why the PR is in draft state). Once I manually test the changes and implement the tests, then I'll mark it ready for review. This was just to make sure that I'm on the right track (before figuring out the tests) 🙂

File: zk/stages/stage_batches.go (hunk 3 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: > With the addition of checking batch number, we need to also check if the batch number of an already processed block matches to the data stream [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/92e7d22534bc103203a4a699b8de60a4a5df4c3e/zk/stages/stage_batches.go#L287-L291).

Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/2a3b9c70438ee4d9d4c3fc963e89e15a1d3bd314

**Comment by hexoscott** [comment]
Conceptually looks good to me

File: zk/stages/stage_batches_test.go [hunk 4]
```diff
@@ -120,3 +96,132 @@ func TestUnwindBatches(t *testing.T) {
...
+
+			dsClient := NewTestDatastreamClient(dsBlocks, nil)
+			for _, l2Block := range dbBlocks {
+				require.NoError(t, hermezDb.WriteBlockBatch(l2Block.L2BlockNumber, l2Block.BatchNumber))
+				require.NoError(t, rawdb.WriteCanonicalHash(tx, l2Block.L2Blockhash, l2Block.L2BlockNumber))
+			}
+			require.NoError(t, tx.Commit())
+
+			erigonDbReader := NewTestErigonDbReader(context.Background(), testDb)
+			hermezDbReader := NewTestHermezDbReaderWrapper(context.Background(), hermez_db.NewHermezDbReader(tx), testDb)
```
Comment (Stefan-Ethernal) [inline zk/stages/stage_batches_test.go]: I've created this wrappers, just because in order to query the MDBX I need to wrap it with  
```go
db.View(ctx, func(tx ...) {
   // querying logic
})
```
However I would be keen to change it if there is a more elegant approach there.

cc @hexoscott @cffls

Comment (hexoscott) [reply zk/stages/stage_batches_test.go]: Hi @Stefan-Ethernal, there are a few things that can help here and some extras:
- The test function name was picked up by the test runner for me until I renamed it to `Test_` with the underscore, or capital `F`.
- When you create the `testDb` and `tx` you should defer the close of the db and the rollback on the tx.
- The commit is the problem here as it closes the transaction down.  If you remove the commit on the transaction you will find the test can just use the `hermezDb.HermezDbReader` without needing the wrapper, and you can use the normal `eridb := erigon_db.NewErigonDb(tx)` for the other case to remove the need for the wrapper.

Comment (Stefan-Ethernal) [reply zk/stages/stage_batches_test.go]: Thanks @hexoscott, this helps! :100: 

> The test function name was picked up by the test runner for me until I renamed it to Test_ with the underscore, or capital F.

I've named it that way, because of this dull Sonarcloud error (https://sonarcloud.io/project/issues?sinceLeakPeriod=true&issueStatuses=OPEN%2CCONFIRMED&pullRequest=886&id=0xPolygonHermez_cdk-erigon&open=AZEnnvAFgHp7PU-smpPO), but if you agree, let's ignore it.

Comment (hexoscott) [reply zk/stages/stage_batches_test.go]: That seems to be a pretty normal way to name test functions to me so happy to ignore the sonar cloud if we can.  If not then just uppercasing the F should make it happy.  Glad the ideas above helped :D

**Review by hexoscott — COMMENTED** [review]
Looking good with the latest changes

File: zk/datastream/client/commands.go [hunk 5]
```diff
@@ -58,6 +58,37 @@ func (c *StreamClient) sendStartCmd(from uint64) error {
...
+	}
+
+	// Send bookmark length
+	if err := writeFullUint32ToConn(c.conn, uint32(len(bookmark))); err != nil {
+		return err
+	}
+
+	// Send the bookmark to retrieve
+	return writeBytesToConn(c.conn, bookmark)
+}
```
Comment (cffls) [inline zk/datastream/client/commands.go]: This is very similar to [sendStartBookmarkCmd](https://github.com/0xPolygonHermez/cdk-erigon/blob/bf724920105f07ee532fa320641435d5845771b4/zk/datastream/client/commands.go#L28-L43). Can we combine the functions into one?

File: zk/stages/stage_batches.go [hunk 6]
```diff
@@ -900,3 +1041,97 @@ func writeL2Block(eriDb ErigonDb, hermezDb HermezDb, l2Block *types.FullL2Block,
...
+	for startBlockNum <= endBlockNum {
+		if endBlockNum == 0 {
+			return 0, emptyHash, ErrFailedToFindCommonAncestor
+		}
+
+		midBlockNum := (startBlockNum + endBlockNum) / 2
+		midBlockDataStream, err := dsClient.GetL2BlockByNumber(midBlockNum)
+		if err != nil &&
+			// the required block might not be in the data stream, so ignore that error
+			!strings.Contains(err.Error(), client.ErrBadFromBookmarkStr) {
```
Comment (cffls) [inline zk/stages/stage_batches.go]: It might be a good idea to check the error code directly, because `ErrBadFromBookmarkStr` may change in a newer data stream version, but error code (4 in this case) is unlikely to change since it is part of the data stream protocol.

File: zk/datastream/client/stream_client.go [hunk 7]
```diff
@@ -86,6 +95,93 @@ func (c *StreamClient) IsVersion3() bool {
 func (c *StreamClient) GetEntryChan() chan interface{} {
 	return c.entryChan
 }
+
+func (c *StreamClient) GetL2BlockByNumber(blockNum uint64) (*types.FullL2Block, error) {
```
Comment (cffls) [inline zk/datastream/client/stream_client.go]: If possible, please have a unit test for this function.

File: zk/datastream/client/stream_client.go [hunk 8]
```diff
@@ -86,6 +95,93 @@ func (c *StreamClient) IsVersion3() bool {
...
+	}
+
+	if l2Block.L2BlockNumber != blockNum {
+		return nil, fmt.Errorf("expected block number %d but got %d", blockNum, l2Block.L2BlockNumber)
+	}
+
+	return l2Block, nil
+}
+
+func (c *StreamClient) GetLatestL2Block() (l2Block *types.FullL2Block, err error) {
```
Comment (cffls) [inline zk/datastream/client/stream_client.go:157]: Same as above. Unit test would be great.

File: zk/datastream/client/stream_client.go [hunk 9]
```diff
@@ -29,15 +30,22 @@ const (
...
+)
+
 type StreamClient struct {
 	ctx          context.Context
 	server       string // Server address to connect IP:port
 	version      int
 	streamType   StreamType
 	conn         net.Conn
-	id           string            // Client id
-	Header       types.HeaderEntry // Header info received (from Header command)
```
Comment (cffls) [inline zk/datastream/client/stream_client.go:39]: Nice! Always good to remove stateful elements.

File: zk/stages/stage_batches.go [hunk 10]
```diff
@@ -76,26 +84,48 @@ type HermezDb interface {
...
+	db                   kv.RwDB
+	blockRoutineStarted  bool
+	dsClient             DatastreamClient
+	dsQueryClientCreator dsClientCreatorHandler
+	zkCfg                *ethconfig.Zk
 }
 
-func StageBatchesCfg(db kv.RwDB, dsClient DatastreamClient, zkCfg *ethconfig.Zk) BatchesCfg {
-	return BatchesCfg{
+func StageBatchesCfg(db kv.RwDB, dsClient DatastreamClient, zkCfg *ethconfig.Zk, options ...Option) BatchesCfg {
```
Comment (cffls) [inline zk/stages/stage_batches.go:100]: I'd prefer using explicit parameter naming here, since it is not a convention to use options as parameters in erigon.

File: zk/stages/stage_batches.go [hunk 11]
```diff
@@ -136,23 +166,53 @@ func SpawnStageBatches(
...
+	if err != nil {
+		log.Warn(fmt.Sprintf("[%s] Error when starting datastream client. Error: %s", logPrefix, err))
+		return err
+	}
+	defer dsQueryClient.Stop()
+
+	highestDSL2Block, err := dsQueryClient.GetLatestL2Block()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve the latest datastream l2 block: %w", err)
+	}
```
Comment (cffls) [inline zk/stages/stage_batches.go]: Let's move this to a function.

File: zk/stages/stage_batches.go [hunk 12]
```diff
@@ -293,7 +354,86 @@ LOOP:
...
+						log.Warn(fmt.Sprintf("[%s] Unwinding to block %d (%s)", logPrefix, unwindBlockNum, unwindBlockHash))
+						err = stages.SaveStageProgress(tx, stages.HighestSeenBatchNumber, batchNum-1)
+						if err != nil {
+							return err
+						}
+						u.UnwindTo(unwindBlockNum, unwindBlockHash)
+
+						cfg.dsClient.Stop()
+
+						return nil
```
Comment (cffls) [inline zk/stages/stage_batches.go]: This is similar to the [code block below](https://github.com/0xPolygonHermez/cdk-erigon/pull/886/files#diff-477089124b274dd75c833b38e783d4cadf9ccaeb6c923e978c2cb092b69a846dR414-R435) where parent where block hash doesn't match. Can we refactor it into a function and use this function in both places?

File: zk/stages/stage_batches.go (hunk 6 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: I wouldn't be worried about error string being changed, because I am using the string from the datastreamer library, by indexing the map of the error codes https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/zk/datastream/client/stream_client.go#L38

In case error string for the `bad from bookmark` gets changed, it is fine, because I'm retrieving the error string that corresponds to the `CmdErrBadFromBookmark` error code.

Your proposal would imply that `GetL2BlockByNumber` needs to return a `ResultEntry` instance (alongside with `error`) and to me it would be more cumbersome for handling from the caller perspective. WDYT?

File: zk/stages/stage_batches.go (hunk 10 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go:100]: It is not the naming here, but instead it is the `Options` pattern and it compensates lack of the constructor/function overloading in the Golang.

In this specific use case, for testing purposes, I need to provide a callback that would create the mock data stream client, i.e. `TestDataStreamClient` (instead of instantiating the one used outside of the tests). However, this parameter should be optional and affect as little codebase as possible.

So in the tests, I'm providing the callback that would allow me to create the mock implementation of DS client https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/zk/stages/stage_batches_test.go#L55. But since this is an optional parameter, the rest of the API is not affected https://github.com/0xPolygonHermez/cdk-erigon/blob/32b3a2926d9b86a9211caff1724d73cc048bc68d/turbo/stages/zk_stages.go#L51 and that's the beauty of this approach.

I wouldn't say I like providing an explicit parameter for the data stream client creator callback function, because it is needed only for testing purposes. However, if you insist on doing so, or have a different proposal, I'm keen to listen out. 🙂 

Resources:
- https://dev.to/kittipat1413/understanding-the-options-pattern-in-go-390c
- https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

File: zk/stages/stage_batches.go (hunk 11 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Do you like this approach https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62?

I didn't want to encapsulate `GetLatestL2Block` as well to the function, because the `dsQueryClient` instance is used for other places as well:
- finding the common ancestor
https://github.com/0xPolygonHermez/cdk-erigon/blob/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62/zk/stages/stage_batches.go#L363
- querying the DS for the parent block hash
https://github.com/0xPolygonHermez/cdk-erigon/blob/4d6b521b342b5b552b9afa2e2d3b824b0ff3fc62/zk/stages/stage_batches.go#L398

So the `newStreamClient` function creates and starts the DS client and that's all (good separation of concerns IMO).

File: zk/datastream/client/commands.go (hunk 5 above)
Comment (Stefan-Ethernal) [reply zk/datastream/client/commands.go]: Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/1017906a099410d8c2032bb98fab2a11b7d41787.

File: zk/stages/stage_batches.go (hunk 12 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/6df39634229ad2adeb2f6157f0b7f46908e1f326

File: zk/datastream/client/stream_client.go (hunk 7 above)
Comment (Stefan-Ethernal) [reply zk/datastream/client/stream_client.go]: It is a bit tricky to be honest, because `net.Pipe` implementation, that is probably the most straightforward approach to mock a network connection blocks on reading and writing (no buffering). `GetL2BlockByNumber` does both writing and reading to the provided connection instance. Do you have some draft idea on how to test it? Not sure whether having a `httptest.Server` would help (https://pkg.go.dev/net/http/httptest#Server), though still don't have a clear vision on how to do it.

File: zk/datastream/client/stream_client.go (hunk 8 above)
Comment (Stefan-Ethernal) [reply zk/datastream/client/stream_client.go:157]: Same as in the https://github.com/0xPolygonHermez/cdk-erigon/pull/886#discussion_r1719886959. Tricky to test, because function uses the same connection to write and read, which is blocking.

File: zk/stages/stage_batches.go (hunk 6 above)
Comment (cffls) [reply zk/stages/stage_batches.go]: Understood the stream is imported from the library, but the problem might occur if the data stream client and the server are using different versions of data stream library. We can't be sure that, at any point of time, the versions of all data stream library used by all clients and servers are always in sync. 
My proposal is to add the error code returned by the server to the error string (if it isn't already there), and check the error code directly.

File: zk/datastream/client/stream_client.go (hunk 7 above)
Comment (cffls) [reply zk/datastream/client/stream_client.go]: Not sure if it is useful, [here](https://github.com/0xPolygonHermez/cdk-erigon/blob/zkevm/zk/da/client_test.go) is an example usage of `httptest.Server`.
Maybe the simplest way is to create a real data stream server and feed some blocks/batches to it in the test? This will not only benefit this PR but for future changes to data stream modules as well.

File: zk/stages/stage_batches.go (hunk 10 above)
Comment (cffls) [reply zk/stages/stage_batches.go:100]: Thanks for the context, it is a beautiful design. I am fine with keeping it.

File: zk/stages/stage_batches.go (hunk 11 above)
Comment (cffls) [reply zk/stages/stage_batches.go]: The whole `SpawnStageBatches` function still looks very complex and does too much than it should. If `dsQueryClient` is reused in other places, maybe pass it as a parameter to the functions that will be refactored out, or create a new ds client in those functions.

Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: > The whole SpawnStageBatches function still looks very complex and does too much than it should. 

I could not agree more with this statement.

> If dsQueryClient is reused in other places, maybe pass it as a parameter to the functions that will be refactored out, or create a new ds client in those functions.

I'm afraid that it would not help to make this function more manageable even with such a change. I can refactor it entirely into some smaller functions, that are easier for understanding. Though is it wise to do this through this PR (it would make the diff even bigger). WDYT?

File: zk/datastream/client/stream_client.go (hunk 8 above)
Comment (Stefan-Ethernal) [reply zk/datastream/client/stream_client.go:157]: After playing around a bit more with `net.Pipe` connection, I've managed to implement a unit test: https://github.com/0xPolygonHermez/cdk-erigon/blob/6b0c1589fba196858efbd54ca13177d70c354768/zk/datastream/client/stream_client_test.go#L272-L410. If you agree I'm about to do the same for `GetL2BlockByNumber`.

Comment (cffls) [reply zk/datastream/client/stream_client.go:157]: This looks great. Thanks! Yes, please have a test for `GetL2BlockByNumber` if possible.

File: zk/stages/stage_batches.go (hunk 6 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/a21e94158cbc17039e153c96c1e087e52a4d159c.

File: zk/datastream/client/stream_client.go (hunk 7 above)
Comment (Stefan-Ethernal) [reply zk/datastream/client/stream_client.go]: Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/0698a90aa40367070732e5a0545df2b628ed689c

File: zk/stages/stage_batches.go (hunk 11 above)
Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Agreed to refactor the entire `SpawnStageBatches` through a separate PR.

**Review by cffls — DISMISSED** [review]
Thanks for addressing all the comments. LGTM!

File: zk/stages/stage_batches.go [hunk 13]
```diff
@@ -288,9 +343,57 @@ LOOP:
...
+
+					if parentBlockDS != nil {
+						dsParentBlockHash = parentBlockDS.L2Blockhash
+					}
+				}
+
+				if dbParentBlockHash != dsParentBlockHash {
+					// unwind/rollback blocks until the latest common ancestor block
+					log.Warn(fmt.Sprintf("[%s] Parent block hashes mismatch on block %d. Triggering unwind...", logPrefix, entry.L2BlockNumber),
+						"db parent block hash", dbParentBlockHash, "ds parent block hash", lastHash)
```
Comment (cffls) [inline zk/stages/stage_batches.go]: Since `lastHash` could be empty, is it better to print `dsParentBlockHash` instead of `lastHash` here?

Comment (Stefan-Ethernal) [reply zk/stages/stage_batches.go]: Good point, agree. Fixed https://github.com/0xPolygonHermez/cdk-erigon/pull/886/commits/dacac355a68f1c9710454ddc21557acb3b9c2d5f

**Review by cffls — DISMISSED** [review]
LGTM

**Review by revitteth — DISMISSED** [review]

**Review by revitteth — DISMISSED** [review]

**Review by revitteth — DISMISSED** [review]

**Review by cffls — APPROVED** [review]
